src/app/modules/ibc.rs

```diff
@@ -399,180 +444,344 @@ impl<S: Store> ConnectionKeeper for Ibc<S> {
-        _port_channel_id: &(PortId, ChannelId),
+        port_channel_id: &(PortId, ChannelId),
     ) -> Result<ChannelEnd, ChannelError> {
-        todo!()
+        let path = IbcPath::ChannelEnds(port_channel_id.0.clone(), port_channel_id.1.clone());
+        let value = self
+            .store
+            .get(Height::Pending, &path.into())
+            .ok_or_else(ChannelError::implementation_specific)?;
+        let channel_end = IbcRawChannelEnd::decode(value.as_slice());
+        channel_end
+            .map_err(|_| ChannelError::implementation_specific())
+            .map(|v| v.try_into().unwrap()) // safety - data on the store is assumed to be well-formed
     }
 
-    fn connection_end(&self, _connection_id: &ConnectionId) -> Result<ConnectionEnd, ChannelError> {
-        todo!()
+    fn connection_end(&self, connection_id: &ConnectionId) -> Result<ConnectionEnd, ChannelError> {
+        ConnectionReader::connection_end(self, connection_id)
+            .map_err(ChannelError::ics03_connection)
     }
 
     fn connection_channels(
         &self,
-        _cid: &ConnectionId,
+        cid: &ConnectionId,
     ) -> Result<Vec<(PortId, ChannelId)>, ChannelError> {
-        todo!()
+        let path = "channelEnds".to_owned().try_into().unwrap(); // safety - path must be valid since ClientId is a valid Identifier
+        let keys = self.store.get_keys(&path);
+        let channels = keys
+            .into_iter()
+            .filter_map(|path| {
+                let value = self.store.get(Height::Pending, &path)?;
+                let channel_end: ChannelEnd = IbcRawChannelEnd::decode(value.as_slice())
+                    .map(|v| v.try_into().unwrap()) // safety - data on the store is assumed to be well-formed
+                    .ok()?;
+
+                if channel_end.connection_hops.first() == Some(cid) {
+                    if let Ok(IbcPath::ChannelEnds(port_id, channel_id)) = path.try_into() {
+                        return Some((port_id, channel_id));
+                    }
+                }
+
+                None
+            })
+            .collect();
+        Ok(channels)
     }
 
-    fn client_state(&self, _client_id: &ClientId) -> Result<AnyClientState, ChannelError> {
-        todo!()
+    fn client_state(&self, client_id: &ClientId) -> Result<AnyClientState, ChannelError> {
+        ConnectionReader::client_state(self, client_id).map_err(ChannelError::ics03_connection)
     }
 
     fn client_consensus_state(
         &self,
-        _client_id: &ClientId,
-        _height: IbcHeight,
+        client_id: &ClientId,
+        height: IbcHeight,
     ) -> Result<AnyConsensusState, ChannelError> {
-        todo!()
+        ConnectionReader::client_consensus_state(self, client_id, height)
+            .map_err(ChannelError::ics03_connection)
     }
 
     fn authenticated_capability(&self, _port_id: &PortId) -> Result<Capability, ChannelError> {
-        todo!()
+        // TODO(hu55a1n1): Copy SDK impl
```

Comment:
For interest's sake, what does this entail? I tried poking around in https://github.com/cosmos/ibc-go but couldn't find the corresponding code (haven't spent a huge amount of time reading through it though, so I probably just missed it).

Reply:
I was thinking we should do something like this -> https://github.com/cosmos/cosmos-sdk/blob/master/x/capability/keeper/keeper.go#L280 
Also, we might need to modify the API on the ibc-rs module side -> https://github.com/informalsystems/ibc-rs/issues/1769